### PR TITLE
userd: generalize dbusInterface

### PR DESCRIPTION
**Pull request by mvo5**
This prepares for having multiple dbus interfaces provided by
snap userd.

Part of the (much) large #4073 but should be trivial and will make the other PR smaller and easier to check.

### CODE REVIEW

File: userd/userd.go [hunk 1]
```diff
@@ -71,17 +50,29 @@ func (ud *Userd) Init() error {
...
 	}
 
-	reply, err := ud.conn.RequestName(busName, dbus.NameFlagDoNotQueue)
-	if err != nil {
-		return err
+	ud.dbusIfaces = []dbusInterface{
+		&Launcher{},
 	}
+	for _, iface := range ud.dbusIfaces {
+		reply, err := ud.conn.RequestName(iface.Name(), dbus.NameFlagDoNotQueue)
```
Comment (bboozzoo) [inline userd/userd.go]: Is each interface going to be exposed as a separate DBus service? Before there was one service with multiple interfaces on a single path. With this change it's service per one of `dbusIfaces`, with multiple interfaces inside.

File: userd/launcher_test.go [hunk 2]
```diff
@@ -28,6 +30,8 @@ import (
 	"github.com/snapcore/snapd/userd"
 )
 
+func Test(t *testing.T) { TestingT(t) }
```
Comment (niemeyer) [inline userd/launcher_test.go]: :flushed:

File: userd/userd.go [hunk 3]
```diff
@@ -71,17 +50,29 @@ func (ud *Userd) Init() error {
...
+			return err
+		}
 
-	if reply != dbus.RequestNameReplyPrimaryOwner {
-		err = fmt.Errorf("cannot obtain bus name '%s'", busName)
-		return err
-	}
+		if reply != dbus.RequestNameReplyPrimaryOwner {
+			err = fmt.Errorf("cannot obtain bus name '%s'", iface.Name())
+			return err
```
Comment (niemeyer) [inline userd/userd.go]: These might be a single line. Nitpick, clearly.. no need to touch it if that's the only change.

File: userd/userd.go [hunk 4]
```diff
@@ -71,17 +50,29 @@ func (ud *Userd) Init() error {
...
+		}
 
-	ud.createAndExportInterfaces()
+		var buffer bytes.Buffer
+		buffer.WriteString("<node>")
+		buffer.WriteString(iface.IntrospectionData())
+		buffer.WriteString(introspect.IntrospectDataString)
+		buffer.WriteString("</node>")
```
Comment (niemeyer) [inline userd/userd.go]: For such simple cases, I suspect this is both simpler and faster than using buffer.String():

`"<node>" + foo + bar "</node>"`

Similar to the other point, this is not a requirement.

File: userd/userd.go (hunk 1 above)
Comment (niemeyer) [reply userd/userd.go]: I suspect this is the main point of the change. We don't want everything under .../Launcher, for example.

File: userd/launcher_test.go (hunk 2 above)
Comment (zyga) [reply userd/launcher_test.go]: Heh, go check could say something like "OMG PANIC" there's a module "foo" that has TestBar function taking a check.C as argument but there's no TestingT integration.

Comment (mvo5) [reply userd/launcher_test.go]: *cough* yes *cough* sorry for this :(

**Pull request merged**